core-common/src/main/java/org/glassfish/jersey/message/internal/ReaderWriter.java

```diff
@@ -157,80 +156,6 @@ public static String readFromAsString(Reader reader) throws IOException {
         }
         return sb.toString();
     }
-    /**
-     * The maximum size of array to allocate.
-     * Some VMs reserve some header words in an array.
-     * Attempts to allocate larger arrays may result in
-     * OutOfMemoryError: Requested array size exceeds VM limit
-     */
-    private static final int MAX_BUFFER_SIZE = Integer.MAX_VALUE - 8;
-
-    /**
-     * Java 9+ InputStream::readAllBytes
-     * TODO Replace when Java 8 not any longer supported (3.1+)
-     */
-    private static byte[] readAllBytes(InputStream inputStream) throws IOException {
```

Comment:
This method is actually good. While it copies the code from the JDK, one can compare the method in JDK 11 and 17, and see some differences. So this method uses the tweaked code from JDK 17.

The method also contains an important fix for streams that have issues with reading 0 bytes from them.

Reply:
I do not see the point. This PR is intended for Jakarta EE 11 which will have JDK 17 or even 21 as its minimum, which brings the code you want by default. So why keeping this method?

Reply:
The change was in the code, these lines:
```
                if (nread == BUFFER_SIZE) { // This differs from JDK version
                    break;                  // prevents a bug (See ReaderWriterTest)
                }
```
The JDK's code reads in a loop until the buffer is full and then the loop asks the stream to read again. That makes the read of zero bytes confuse some streams and return -1. While the stream's behavior is arguably incorrect, it would be a sudden change in Jersey that would make many years of running code fail miserably for a large customer. 

Is there a strong reason for dropping the code?

Reply:
Yes, there is a strong reason: Future JRE-internal optimizations will not be effective in Jersey (e. g. hypothetically we might decide to replace Java completely by C code).

Certainly the investments done in OpenJDK SHOULD (MUST?) reflect its benefits to Jersey users, too. So that situation would be a major drawback for *all* Jersey users! In an open source project it simply MUST NOT happen that one single party (read: customer) prevents *all other* parties from progressing. I think we all agree on this, particularly now that we are hosted at the EF.

BTW, the same problem could arise (for the same customer and any other users) at any time in *any other* application, as it is not limited to Jersey, but limited to the particular stream implementation. So if such a fix is *essentially needed*, it has to go into the JRE, to be on the safe side, instead of limiting the fix particularly to Jersey. But if the fix is *not needed* (hence was only a nice gift to the customer), the customer would be really on the safer side when he fixes his custom stream instead (which I assume, but don't know, to be the case here).

Having said that, a compromise could be:
* Let's keep the code *for 3.x* so the customer / we gain more time for a real fix *at the right place*.
* Let me check with my OpenJDK peers if we add the fix to JDK 21 / 22. If they refuse to accept the change proposal, the customer should get contacted that the nice gift will go away eventually.
* Let's drop the code in Jersey 4.

I think that would be fair to *all* parties.

Reply:
Yep, keeping it for 3.x sounds like a good compromise. It could have been revisited once the JDK really brings in an internal optimization. Right now, keeping the method as optimized as possible from the newest JDK would bring positives to people using old Java (11).

Reply:
Per our compromise I reintroduced the original code in 6b5d1948d0ea8f147b1bf01c65f6b20ef0386388, so the workaround still exists until Jersey 4. I have updated the TODO so nobody gets confused by the original justification. I also filed [an issue](https://github.com/eclipse-ee4j/jersey/issues/5344) so we do not forget to fix this in Jersey 4.

JDK 21 is already in Rampdown Phase, so *even if* Team OpenJDK adops the workaround (which is not decided yet) it will not make it into any JDK earlier than 22. It might make sense to contact the customer right now, as chances to get the workaround into JDK are actually rather low.